protocols/perf/src/client/handler.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright 2023 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::{
+    collections::VecDeque,
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+
+use futures::{future::BoxFuture, stream::FuturesUnordered, FutureExt, StreamExt, TryFutureExt};
+use libp2p_core::upgrade::{DeniedUpgrade, ReadyUpgrade};
+use libp2p_swarm::{
+    handler::{
+        ConnectionEvent, DialUpgradeError, FullyNegotiatedInbound, FullyNegotiatedOutbound,
+        ListenUpgradeError,
+    },
+    ConnectionHandler, ConnectionHandlerEvent, ConnectionHandlerUpgrErr, KeepAlive,
+    SubstreamProtocol,
+};
+use void::Void;
+
+use super::{RunId, RunParams, RunStats};
+
+#[derive(Debug)]
+pub struct Command {
+    pub(crate) id: RunId,
+    pub(crate) params: RunParams,
+}
+
+#[derive(Debug)]
+pub struct Event {
+    pub(crate) id: RunId,
+    pub(crate) result: Result<RunStats, ConnectionHandlerUpgrErr<Void>>,
+}
+
+pub struct Handler {
+    /// Queue of events to return when polled.
+    queued_events: VecDeque<
+        ConnectionHandlerEvent<
+            <Self as ConnectionHandler>::OutboundProtocol,
+            <Self as ConnectionHandler>::OutboundOpenInfo,
+            <Self as ConnectionHandler>::OutEvent,
+            <Self as ConnectionHandler>::Error,
+        >,
+    >,
+
+    outbound: FuturesUnordered<BoxFuture<'static, Result<Event, std::io::Error>>>,
+
+    keep_alive: KeepAlive,
+}
+
+impl Handler {
+    pub fn new() -> Self {
+        Self {
+            queued_events: Default::default(),
+            outbound: Default::default(),
+            keep_alive: KeepAlive::Yes,
+        }
+    }
+}
+
+impl Default for Handler {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ConnectionHandler for Handler {
+    type InEvent = Command;
+    type OutEvent = Event;
+    type Error = std::io::Error;
```

Comment:
```suggestion
    type Error = Void;
```
 You never close the connection actively.

Reply:
cd8baf6e